package java17.ex01;

import static org.hamcrest.Matchers.anyOf;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.Matchers.everyItem;
import static org.hamcrest.Matchers.greaterThan;
import static org.hamcrest.Matchers.hasProperty;
import static org.hamcrest.Matchers.hasSize;
import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import org.junit.Test;

import java17.data.Data;
import java17.data.domain.Order;
import java17.data.domain.Pizza;

/**
 * Exercice 01 - Recherche
 */
public class Stream_01_Test {

	@Test
	public void test_stream_filter() throws Exception {
		List<Pizza> pizzas = new Data().getPizzas();

		// TODO récupérer la liste des pizzas dont le prix est >= 1300

		List<Pizza> pizzasCheres = pizzas.stream()
				.filter(pizza -> pizza.getPrice() >= 1300)
				.collect(Collectors.toList());

		assertThat(pizzasCheres, hasSize(3));
		assertThat(pizzasCheres, everyItem(hasProperty("price", anyOf(equalTo(1300), greaterThan(1300)))));
	}

	@Test
	public void test_stream_filter_collect_counting() throws Exception {
		List<Pizza> pizzas = new Data().getPizzas();

		// TODO compter le nombre de pizzas dont le prix est >= 1300
		long nbPizzasCheres = pizzas.stream()
				.filter(pizza -> pizza.getPrice() >= 1300)
				.count();

		assertThat(nbPizzasCheres, is(3L));
	}

	@Test
	public void test_stream_anyMatch() throws Exception {

		List<Pizza> pizzas = new Data().getPizzas();

		// TODO valider si au moins une pizza à un prix >= 1300
		boolean aUnePizzaChere = pizzas.stream()
				.anyMatch(pizza -> pizza.getPrice() >= 1300);

		// TODO valider si au moins une pizza à un prix >= 2000
		boolean aUnePizzaChere1 = pizzas.stream()
				.anyMatch(pizza -> pizza.getPrice() >= 2000);

		assertThat(aUnePizzaChere, is(true));
		assertThat(aUnePizzaChere1, is(false));
	}

	@Test
	public void test_stream_allMatch() throws Exception {

		List<Pizza> pizzas = new Data().getPizzas();

		// TODO valider que toutes les pizzas ont un prix >= 1300
		boolean toutesLesPizzasSontCheres = pizzas.stream()
				.allMatch(pizza -> pizza.getPrice() >= 1300);

		// TODO valider que toutes les pizzas ont un prix >= 900
		boolean toutesLesPizzasSontCheres1 = pizzas.stream()
				.allMatch(pizza -> pizza.getPrice() >= 900);

		assertThat(toutesLesPizzasSontCheres, is(false));
		assertThat(toutesLesPizzasSontCheres1, is(true));
	}

	@Test
	public void test_stream_noneMatch() throws Exception {

		List<Pizza> pizzas = new Data().getPizzas();

		// TODO valider qu'aucune pizza n'a un prix >= 2000
		boolean aucunePizzaChere = pizzas.stream().noneMatch(pizza -> pizza.getPrice() >= 2000);
		assertThat(aucunePizzaChere, is(true));
	}

	@Test
	public void test_stream_findFirst() throws Exception {
		List<Order> orders = new Data().getOrders();

		// TODO récupérer une commande faite par un client dont le prénom est "Sophie"
		Optional<Order> order = orders.stream()
				.filter(o -> o.getCustomer().getFirstname().equalsIgnoreCase("Sophie"))
				.findFirst();

		assertThat(order.isPresent(), is(false));
	}

	/**
	 * FACULTATIVE : PLUS DURE !!
	 */
	@Test
	public void test_stream_filter_adv() throws Exception {
		List<Order> orders = new Data().getOrders();

		// TODO récupérer la liste de toutes les commandes qui contiennent au moins une
		// pizza Pépéroni
		List<Order> result = orders.stream()
				.filter(order -> order.getPizzas().stream().anyMatch(item -> "Pepperoni".equalsIgnoreCase(item.getName())))
				.toList();

		assertThat(result, hasSize(3));
	}
}
